packages/mwp-router/src/AsyncRoute.jsx

```diff
@@ -10,90 +10,86 @@ type Props = {
 	location: URL,
 	history: RouterHistory,
 };
-type State = {
-	childRoutes: Array<PlatformRoute>,
-	_routesCache: { [string]: Array<PlatformRoute> },
+type ComponentState = {
+	component: React$ComponentType<*>,
+	_componentCache: { [string]: React$ComponentType<*> },
 };
+type State = ComponentState;
+
+// simple pass through component to use while real component is loading
+const PassThrough = (children: React$Node) => React.Children.only(children);
```

Comment:
I'm interested to see how this plays in a real application - it effectively allows 'inner' routes/containers to render _before_ their parent. For the most part, this won't matter, but there may be cases where a parent provides some wrapping chrome that will pop in after a delay - should be rare, and I think that's preferable to _not_ rendering the child content (which is what the people came to see) if it's ready before the parent wrapper.

Reply:
Yeah I want to see it in action as well. There are a few things that should mitigate the likelihood of that kind of layout thrashing, however.

1. _Most_ of the time people will be navigating from a view that has already loaded the 'wrapper' component. Not always, but most of the time.
2. When someone first loads the app on a particular page, the client will have pre-loaded all the matching components, so the layout shouldn't thrash in that case.
3. The service worker will greedily download all components for _unmatched_ routes when the app first loads, so by the time someone navigates within the app, it's very likely that the new route will already have access to all the components it needs.

Still need to see it in action to confirm, however. I'm just realizing that there might be some issue with the initial render of the client-side app not matching the server markup, which might require some tweaks to avoid warnings, but we'll see.

Reply:
sounds good to me :)